### ISSUE TITLE
ArcFace: Additive Angular Margin Loss for Deep Face Recognition

### ISSUE THREAD

**Issue by agatan** [agatan/paper-reading-notes#4]
- https://arxiv.org/abs/1801.07698

face recognition のような open-set なクラス分類タスクのための枠組みの提案。
ざっくり triplet loss を置き換えるものだと思ったら良い。

ふつうのクラス分類は softmax をつかうが、softmax はクラス数に応じて計算量やメモリ使用量が増えるし、closed-set なタスクにしか使えない。

facenet で使われている triplet loss は anchor, positive,  negative の 3 つ組をたくさん作り `anchor ↔ positive + margin < anchor ↔ negative` になるように学習する。
triplet loss の問題は、↑のような 3 つ組はめちゃくちゃ大量に考えられるのでデータ量が爆発するし hard example を見つけるのが大変すぎる。

そこで、ふつうのクラス分類タスクとして解きつつ各クラス間に十分なマージンができるように学習することで、学習を簡単にしつつ open-set なクラス分類を実現する。
具体的には softmax + crossentropy の代わりに以下のような損失関数を使う。

![image](https://user-images.githubusercontent.com/6925889/54260961-c64b6780-45ad-11e9-9889-284f6d4e63c1.png)

ここで、 `cos θ` は各クラスの代表ベクトル `W` と入力画像の embedding `x` の間の角度。
